acme/acme/client.py

```diff
@@ -837,16 +840,18 @@ def _assess_tos(tos):
             if check_tos_cb is not None:
                 check_tos_cb(tos)
         if self.acme_version == 1:
-            regr = self.client.register(regr)
+            client_v1 = cast(Client, self.client)
```

Comment:
Yes I understand, I also hesitate between several solutions. The naming of this function is not a great move, in my opinion, because it remind us (developpers) too much of a real cast, which is never something you feel totally comfortable to do in the first place, and frequently is a synonym of some bad design in your code...

Here as the doc states, `typing.cast` will do nothing at runtime, it is really only for mypy and help it understand the type when the code cannot let it infer that. Basically here, we know from `__init__` that if `self.acme_version ==1`, then `self.client` is of type `Client`, and `ClientV2` otherwise.

I thought about two other ways to fix the problem here, but each of these approaches was not better in my opinion:
* the most obvious, `if isinstance(self.client, Client)`: it is elegant on mypy side, because alonside with the fact that `self.client is declared` as `Union[Client, ClientV2`, it will properly infer this case (and also my IDE) when `self.client` is actually a `Client` or a `ClientV2`. Sadly tests are not happy because we frequently mock these classes, making the `isinstance` not working as expected. For sure I could use a `spec` on the `MagikMock`, but at that time it was too much code compared to a simple `cast`.
* as an alternative, assign the ACME client to `self.client_v1` or `self.client_v2`, then use `if self.client_v1` or `if self.client_v2`. First, it adds a lot of code to the `BackwardCompatibleClient` juste to make mypy happy. Also, you still need to have two `if`, and not a `if/else` because both `self.client_v1` and `self.client_v2` will be `Optionals[]`. All of this for a class which is fundamentally planned to be removed at some point.

Reply:
Thanks for the write up. I personally think what you have here is fine. I was mainly just expressing my dissatisfaction about how limited our options are here and the amount of work `mypy` requires from us to be satisfied when we know this code is OK. In general, I'm slightly allergic to spending much time adapting our work to satisfy tools instead the tools adapting themselves around our practices. At the same time, I do think we should probably always at least consider refactoring our code to make `mypy` happy instead of silencing them, but I don't think it's worth it here.